### PR TITLE
VarExpand for the register runtime

### PR DESCRIPTION
**Pull request by systay**

### CODE REVIEW

**Comment by systay** [comment]
This is built on top of #9759 to show which scenarios are successful after this change

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala [hunk 1]
```diff
@@ -41,6 +41,21 @@ class MatchAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsTe
     result.toList should equal(List(Map("count(n.name)" -> 3)))
   }
 
+  test("varlength apa") {
```
Comment (craigtaverner) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Seems like a non-ideal test name....

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocation.scala [hunk 2]
```diff
@@ -141,6 +141,12 @@ object RegisterAllocation {
         result += (lp -> cartesianProductPipeline)
         cartesianProductPipeline
 
+      case VarExpand(lhs, IdName(from), dir, projectedDir, types, IdName(to), IdName(edge), length, ExpandAll, predicates) =>
+        val pipeline = allocate(lhs, nullable, argument)
```
Comment (craigtaverner) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocation.scala]: Why doesn't this do the same as expand and create a new pipe using deepClone? Since the VarExpand, like the normal Expand, will change the number of columns, we would expect a new pipeline, right?

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/RegisteredPipeBuilder.scala [hunk 3]
```diff
@@ -94,6 +95,22 @@ class RegisteredPipeBuilder(fallback: PipeBuilder,
         val toSlot = pipeline(to)
         ExpandIntoRegisterPipe(source, fromSlot.offset, relSlot.offset, toSlot.offset, dir, LazyTypes(types), pipeline)(id)
 
+      case VarExpand(_, IdName(fromName), dir, projectedDir, types, IdName(toName), IdName(relName), VarPatternLength(min, max), expansionMode, predicates) =>
+        // TODO: This is not right!
+        if(predicates.nonEmpty)
+          throw new CantCompileQueryException("does not handle varexpand with predicates")
```
Comment (craigtaverner) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/RegisteredPipeBuilder.scala:123]: Other PR's handing OptionalExpandRegisterPipe have solutions for predicates, so we should reuse those.

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/VarLengthExpandRegisterPipe.scala [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      inputRowWithFromNode =>
+        val fromNode = inputRowWithFromNode.getLongAt(fromOffset)
+        val paths: Iterator[(LNode, Seq[LRelationship])] = varLengthExpand(fromNode, state, inputRowWithFromNode)
+        paths collect {
+          case (toNode: LNode, relIds: Seq[LRelationship])
+            if relIds.length >= min && isToNodeValid(inputRowWithFromNode, toNode) =>
+            val resultRow = PrimitiveExecutionContext(pipeline)
+            resultRow.copyFrom(inputRowWithFromNode)
+            resultRow.setLongAt(toOffset, toNode)
+            val rels = relIds.map(state.query.relationshipOps.getById)
```
Comment (craigtaverner) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/VarLengthExpandRegisterPipe.scala]: Not great to have to hit the store to re-lookup the relationships again, when we know the previous lookup already got all this information.

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/VarLengthExpandRegisterPipe.scala [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            resultRow.setRefAt(relOffset, rels)
+            resultRow
+        }
+    }
+  }
+
+  private def isToNodeValid(row: ExecutionContext, node: LNode): Boolean =
+    !closedPath || row.getLongAt(toOffset) == node
```
Comment (craigtaverner) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/VarLengthExpandRegisterPipe.scala]: The boolean closedPath is non-intuitive. Would be better with something like 'endNodeBound', or even 'ExpandInto'

**Review by craigtaverner — CHANGES_REQUESTED** [review]

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/LernaeanTestSupport.scala [hunk 6]
```diff
@@ -87,7 +87,7 @@ trait LernaeanTestSupport extends CypherTestSupport {
   val newRuntimeMonitor = new NewRuntimeMonitor
 
   private def extractFirstScenario(config: TestConfiguration): TestScenario = {
-    val preferredScenario = Scenarios.CommunityInterpreted
+    val preferredScenario = Scenarios.CommunityInterpreted3_3
```
Comment (craigtaverner) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/LernaeanTestSupport.scala]: This seems to be unrelated to this work, and also seems incomplete, because the same improvement is not done to EnterpriseInterpreted or Compiled.

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala (hunk 1 above)
Comment (craigtaverner) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Perhaps this was a development-only test? Not meant to be committed?

**Review by craigtaverner — CHANGES_REQUESTED** [review]
I did not find any tests for this work. Presumably at least the RegisterPipeBuilderTest should be enhanced to cover this?

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala (hunk 1 above)
Comment (systay) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Yep

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocation.scala (hunk 2 above)
Comment (systay) [reply enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/RegisterAllocation.scala]: Absolutely right

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/RegisteredPipeBuilder.scala (hunk 3 above)
Comment (systay) [reply enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/RegisteredPipeBuilder.scala:123]: Very different. I have been working on solving this, but it can be done as a separate step. I'll push what I have so you can see what I'm thinking about.

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/VarLengthExpandRegisterPipe.scala (hunk 4 above)
Comment (systay) [reply enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/pipes/VarLengthExpandRegisterPipe.scala]: +1

**Review by craigtaverner — APPROVED** [review]
